sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/implementation/PartitionBasedLoadBalancer.java

```diff
@@ -0,0 +1,353 @@
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Random;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import reactor.core.Exceptions;
+import reactor.core.publisher.Mono;
+import reactor.util.function.Tuple2;
+
+/**
+ * This class is responsible for balancing the load of processing events from all partitions of an Event Hub by
+ * distributing the number of partitions uniformly among all the  active {@link EventProcessor EventProcessors}.
+ * <p>
+ * This load balancer will retrieve partition ownership details from the {@link PartitionManager} to find the number of
+ * active {@link EventProcessor EventProcessors}. It uses the last modified time to decide if an EventProcessor is active. If a
+ * partition ownership entry has not be updated for a specified duration of time, the owner of that partition is
+ * considered inactive and the partition is available for other EventProcessors to own.
+ * </p>
+ */
+public final class PartitionBasedLoadBalancer {
+
+    private static final Random RANDOM = new Random();
+    private final ClientLogger logger = new ClientLogger(PartitionBasedLoadBalancer.class);
+
+    private final String eventHubName;
+    private final String consumerGroupName;
+    private final PartitionManager partitionManager;
+    private final EventHubAsyncClient eventHubAsyncClient;
+    private final String ownerId;
+    private final long inactiveTimeLimitInSeconds;
+    private final PartitionPumpManager partitionPumpManager;
+
+    /**
+     * Creates an instance of PartitionBasedLoadBalancer for the given Event Hub name and consumer group.
+     *
+     * @param partitionManager The partition manager that this load balancer will use to read/update ownership details.
+     * @param eventHubAsyncClient The asynchronous Event Hub client used to consume events.
+     * @param eventHubName The Event Hub name the {@link EventProcessor} is associated with.
+     * @param consumerGroupName The consumer group name the {@link EventProcessor} is associated with.
+     * @param ownerId The identifier of the {@link EventProcessor} that owns this load balancer.
+     * @param inactiveTimeLimitInSeconds The time in seconds to wait for an update on an ownership record before
+     * assuming the owner of the partition is inactive.
+     * @param partitionPumpManager The partition pump manager that keeps track of all EventHubConsumers and partitions
+     * that this {@link EventProcessor} is processing.
+     */
+    public PartitionBasedLoadBalancer(final PartitionManager partitionManager,
+        final EventHubAsyncClient eventHubAsyncClient,
+        final String eventHubName, final String consumerGroupName, final String ownerId,
+        final long inactiveTimeLimitInSeconds, final PartitionPumpManager partitionPumpManager) {
```

Comment:
`inactiveTimeLimitInSeconds` I wonder if it would be easier to understand if we used a class like Duration rather than a long.